dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp

```diff
@@ -199,51 +201,50 @@ ScanHashMapAfterProbeBlockInputStream::ScanHashMapAfterProbeBlockInputStream(
           * result_sample_block - keys, "left" columns, and "right" columns.
           */
 
-    size_t num_columns_left = left_sample_block.columns();
-    if (isRightSemiFamily(parent.getKind()))
-        num_columns_left = 0;
-    else
-        result_sample_block = materializeBlock(left_sample_block);
-
-    size_t num_columns_right = parent.sample_block_with_columns_to_add.columns();
-    /// Add columns from the right-side table to the block.
-    for (size_t i = 0; i < num_columns_right; ++i)
+    column_indices_left.reserve(left_sample_block.columns());
+    if (!isRightSemiFamily(parent.getKind()))
     {
-        const ColumnWithTypeAndName & src_column = parent.sample_block_with_columns_to_add.getByPosition(i);
-        result_sample_block.insert(src_column.cloneEmpty());
+        auto left_full_block = materializeBlock(left_sample_block);
+        for (size_t i = 0; i < left_full_block.columns(); ++i)
+        {
+            auto & column = left_full_block.getByPosition(i);
+            if (parent.output_column_names_set_after_finalize.contains(column.name))
+            {
+                result_sample_block.insert(column.cloneEmpty());
+                column_indices_left.push_back(i);
+            }
+        }
     }
 
-    column_indices_left.reserve(num_columns_left);
-    column_indices_right.reserve(num_columns_right);
-
-    for (size_t i = 0; i < num_columns_left; ++i)
+    column_indices_right.reserve(parent.sample_block_without_keys.columns());
+    /// Add columns from the right-side table to the block.
+    for (size_t i = 0; i < parent.sample_block_without_keys.columns(); ++i)
     {
-        column_indices_left.push_back(i);
+        const ColumnWithTypeAndName & src_column = parent.sample_block_without_keys.getByPosition(i);
+        if (parent.output_column_names_set_after_finalize.contains(src_column.name))
+        {
+            result_sample_block.insert(src_column.cloneEmpty());
+            column_indices_right.push_back(i);
+        }
     }
 
-    for (size_t i = 0; i < num_columns_right; ++i)
-        column_indices_right.push_back(num_columns_left + i);
-
-    for (size_t i = 0; i < num_columns_left; ++i)
+    for (size_t i = 0; i < column_indices_left.size(); ++i)
     {
-        const auto & column_with_type_and_name = result_sample_block.getByPosition(column_indices_left[i]);
+        const auto & column_with_type_and_name = result_sample_block.getByPosition(i);
```

Comment:
why change here?

Reply:
Before this change, `column_indices_left[i]` is always `i`, and `i` is also the column pos in both `result_sample_block` and `left_sample_block`
After this change, `column_indices_left[i]` is not always `i`, and `column_indices_left[i]` is the column pos in `left_sample_block`, `i` is the column pos in `result_sample_block`

Reply:
ok